src/Microsoft.Diagnostics.EventFlow.Core/Implementations/LossReportingPropagatorBlock.cs

```diff
@@ -0,0 +1,209 @@
+        {
+            Requires.NotNull(target, nameof(target));
+            if (this.target != null)
+            {
+                throw new InvalidOperationException($"{nameof(LossReportingPropagatorBlock<T>)} only supports a single target");
+            }
+
+            this.target = target;
+            if (linkOptions != null)
+            {
+                // Note DataflowLinkOptions.Append does not matter given that we are working with a single target only 
+                // and we will not enforce DataflowLinkOptions.MaxMessages
+                this.propagateCompletion = linkOptions.PropagateCompletion;
+            }
+
+            // We do not support re-linking
+            return new EmptyDisposable();
+        }
+
+        public DataflowMessageStatus OfferMessage(DataflowMessageHeader messageHeader, T messageValue, ISourceBlock<T> source, bool consumeToAccept)
+        {
+            VerifySourceInitiatedOperation(source);
+
+            if (this.completed)
+            {
+                return DataflowMessageStatus.DecliningPermanently;
+            }
+
+            var targetResponse = this.target.OfferMessage(messageHeader, messageValue, this, consumeToAccept);
+            switch(targetResponse)
+            {
+                case DataflowMessageStatus.Accepted:
+                    // Great, nothing to do
+                    break;
+
+                case DataflowMessageStatus.Declined:
+                case DataflowMessageStatus.DecliningPermanently:
+                case DataflowMessageStatus.NotAvailable:
+                    this.healthReporter.ReportThrottling();
+                    break;
+
+                case DataflowMessageStatus.Postponed:
+                    // This is a bit tricky--the message might be successfully consumed later, or it might no longer be available.
+                    // We'll know for sure based on the success/failure of ConsumeMessage().
+                    break;
+            }
+
+            return targetResponse;
+        }
+
+        public T ConsumeMessage(DataflowMessageHeader messageHeader, ITargetBlock<T> target, out bool messageConsumed)
+        {
+            VerifyTargetInitiatedOperation(target);
+
+            messageConsumed = false;
+
+            if (this.completed)
+            {
+                return default(T);
+            }
+
+            T message = this.source.ConsumeMessage(messageHeader, this, out bool consumedByMe);
+            if (consumedByMe)
+            {
+                messageConsumed = true;
+                return message;
+            }
+            else
+            {
+                // Message is no longer available
+                this.healthReporter.ReportThrottling();
+                return default(T);
+            }
+        }
+
+        public void ReleaseReservation(DataflowMessageHeader messageHeader, ITargetBlock<T> target)
+        {
+            VerifyTargetInitiatedOperation(target);
```

Comment:
It feels like this Reserve/Release APIs will and should never be used on this class.